pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -261,17 +259,22 @@ func BuildHTTPRoutesForVirtualService(
 	vsName := virtualService.ConfigMeta.Name
 
 	out := make([]route.Route, 0, len(vs.Http))
+allroutes:
 	for _, http := range vs.Http {
 		if len(http.Match) == 0 {
 			if r := translateRoute(push, node, http, nil, port, vsName, serviceRegistry, proxyLabels, gatewayNames); r != nil {
 				out = append(out, *r)
 			}
-			break // we have a rule with catch all match prefix: /. Other rules are of no use
+			break allroutes // we have a rule with catch all match prefix: /. Other rules are of no use
 		} else {
-			// TODO: https://github.com/istio/istio/issues/4239
 			for _, match := range http.Match {
 				if r := translateRoute(push, node, http, match, port, vsName, serviceRegistry, proxyLabels, gatewayNames); r != nil {
 					out = append(out, *r)
+					rType, _ := getEnvoyRouteTypeAndVal(r)
+					if rType == envoyCatchAll {
+						// We have a catch all route. No point building other routes, with match conditions
+						break allroutes
+					}
```

Comment:
Should we add something to pushcontext - if the user has other routes after, it's likely a bad config.

Reply:
Or we could update the doc and change behavior - so default route is applied last.

Reply:
I think adding to push context would be a lot ? If anything, we should catch this in validation and warn.